File: Part1_강의자료/Ch08_투포인터/문제별코드/2118_두개의탑/2118.java
import java.util.Scanner;

class Main
{
    public static void main (String[] args) {
        Scanner sc = new Scanner(System.in);

        int N = sc.nextInt();
        int[] distance = new int[N];
        int distanceSum = 0;
        for (int i = 0; i < N; i++) {
            distance[i] = sc.nextInt();
            distanceSum += distance[i];
        }

        int pairIndex = 1;
        int leftDistance = distance[0];
        int rightDistance = distanceSum - distance[0];
        int maxDistance = Math.min(leftDistance, rightDistance);
        for (int i = 0; i < N; i++) {
            while (leftDistance < rightDistance) {
                leftDistance += distance[pairIndex];
                rightDistance -= distance[pairIndex];
                pairIndex = (pairIndex + 1) % N;
            }
            maxDistance = Math.max(maxDistance, Math.min(leftDistance, rightDistance));

            leftDistance -= distance[i];
            rightDistance += distance[i];
        }
        System.out.println(maxDistance);
    }
}

File: Part1_강의자료/Ch08_투포인터/문제별코드/2118_두개의탑/2118.cpp
#include <stdio.h>
#include <vector>
#include <algorithm>
using namespace std;

int main()
{
    int N;
    
    scanf("%d", &N);
    vector<int> distance(N);
    int distance_sum = 0;
    for (int i = 0; i < N; i++)
    {
        scanf("%d", &distance[i]);
        distance_sum += distance[i];
    }

    int pair_index = 1;
    int left_distance = distance[0];
    int right_distance = distance_sum - distance[0];
    int max_distance = min(left_distance, right_distance);
    for (int i = 0; i < N; i++)
    {
        while (left_distance < right_distance)
        {
            left_distance += distance[pair_index];
            right_distance -= distance[pair_index];
            pair_index = (pair_index + 1) % N;
        }
        max_distance = max(max_distance, min(left_distance, right_distance));

        left_distance -= distance[i];
        right_distance += distance[i];
    }

    printf("%d\n", max_distance);
    return 0;
}
